File: src/islam/adhanalarm/view/QiblaCompassView.java
package islam.adhanalarm.view;

import java.text.DecimalFormat;

import uz.efir.muazzin.R;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

public class QiblaCompassView extends View {
    private float directionNorth = 0;
    private float directionQibla = 0;
    private TextView bearingNorth;
    private String bearingNorthString;
    private TextView bearingQibla;
    private String bearingQiblaString;
    private DecimalFormat df = new DecimalFormat("0.000");
    private Bitmap compassBackground;
    private Bitmap compassNeedle;
    private Matrix rotateNeedle = new Matrix();
    private int width = 240;
    private int height = 240;
    private float centre_x = width * 0.5f;
    private float centre_y = height * 0.5f;

    public QiblaCompassView(Context context) {
        super(context);
        initCompassView();
    }
    public QiblaCompassView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initCompassView();
    }
    public QiblaCompassView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initCompassView();
    }

    @Override
    public void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(width, height);
    }

    private void initCompassView() {
        compassNeedle = BitmapFactory.decodeResource(getResources(), R.drawable.compass_needle);
        compassBackground = BitmapFactory.decodeResource(getResources(), R.drawable.compass_background);
        width = getWidth(compassBackground, 240);
        height = getHeight(compassBackground, 240);
        centre_x = width  * 0.5f;
        centre_y = height * 0.5f;
        rotateNeedle.postTranslate(centre_x - getWidth(compassNeedle, 23) + 10, centre_y - getHeight(compassNeedle, 142) + 10);
        invalidate();
    }

    public void setConstants(TextView bearingNorth, CharSequence bearingNorthString, TextView bearingQibla, CharSequence bearingQiblaString) {
        this.bearingNorth = bearingNorth;
        this.bearingNorthString = bearingNorthString.toString();
        this.bearingQibla = bearingQibla;
        this.bearingQiblaString = bearingQiblaString.toString();
        compassBackground = BitmapFactory.decodeResource(getResources(), R.drawable.compass_background);
        compassNeedle = BitmapFactory.decodeResource(getResources(), R.drawable.compass_needle);
    }

    public void setDirections(float directionNorth, float directionQibla) {
        this.directionNorth = directionNorth;
        this.directionQibla = directionQibla;
        rotateNeedle = new Matrix();
        rotateNeedle.postRotate(-directionQibla, getWidth(compassNeedle, 23) * 0.5f, getHeight(compassNeedle, 142));
        rotateNeedle.postTranslate(centre_x - getWidth(compassNeedle, 23), centre_y - getHeight(compassNeedle, 142) + 5);
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        bearingNorth.setText(bearingNorthString.replace("(1)", df.format(directionNorth)));
        bearingQibla.setText(bearingQiblaString.replace("(+/-)", directionQibla >= 0 ? " +" : " -").replace("(2)", df.format(Math.abs(directionQibla))).replace("(3)",  df.format(directionNorth + directionQibla)));

        Paint p = new Paint();
        canvas.rotate(-directionNorth, centre_x, centre_y);
        canvas.drawBitmap(compassBackground, 0, 0, p);
        canvas.drawBitmap(compassNeedle, rotateNeedle, p);
    }

    private int getHeight(Bitmap bitmap, int def) {
        if (isInEditMode()) {
            return def;
        }

        return bitmap.getHeight();
    }

    private int getWidth(Bitmap bitmap, int def) {
        if (isInEditMode()) {
            return def;
        }

        return bitmap.getWidth();
    }
}

File: app/src/main/kotlin/uz/efir/muazzin/QiblaCompassView.kt
package uz.efir.muazzin

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.graphics.withRotation

class QiblaCompassView : View {
    private var directionNorth = 0f
    private var directionQibla = 0f
    private val compassBackground: Drawable
    private val compassNeedle: Drawable
    private val labelPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
        typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
    }

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
        context, attrs, defStyle
    ) {
        labelPaint.color = ContextCompat.getColor(context, R.color.text_primary)

        compassBackground = requireNotNull(
            ContextCompat.getDrawable(context, R.drawable.compass_background)
        )
        compassNeedle = requireNotNull(
            ContextCompat.getDrawable(context, R.drawable.compass_needle)
        )
    }

    public override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val defaultSizePx = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, DEFAULT_SIZE_DP, resources.displayMetrics
        ).toInt()
        val w = resolveSize(defaultSizePx, widthMeasureSpec)
        val h = resolveSize(defaultSizePx, heightMeasureSpec)
        val side = minOf(w, h)
        setMeasuredDimension(side, side)
    }

    fun setDirections(directionNorth: Float, directionQibla: Float) {
        this.directionNorth = directionNorth
        this.directionQibla = directionQibla
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        val side = width
        val centerX = side / 2f
        val centerY = side / 2f

        val compassSide = side * 0.8f
        val compassOffset = (side - compassSide) / 2f

        canvas.save()
        canvas.rotate(-directionNorth, centerX, centerY)

        compassBackground.setBounds(
            compassOffset.toInt(),
            compassOffset.toInt(),
            (compassOffset + compassSide).toInt(),
            (compassOffset + compassSide).toInt()
        )
        compassBackground.draw(canvas)

        labelPaint.textSize = side * LABEL_TEXT_RATIO
        val labelRadius = side * LABEL_RADIUS_RATIO
        val baselineOffset = labelPaint.textSize * 0.35f

        val labels = arrayOf("N", "E", "S", "W")
        for (i in labels.indices) {
            canvas.withRotation(i * 90f, centerX, centerY) {
                drawText(labels[i], centerX, centerY - labelRadius + baselineOffset, labelPaint)
            }
        }

        canvas.save()
        canvas.rotate(directionQibla, centerX, centerY)
        compassNeedle.setBounds(
            compassOffset.toInt(),
            compassOffset.toInt(),
            (compassOffset + compassSide).toInt(),
            (compassOffset + compassSide).toInt()
        )
        compassNeedle.draw(canvas)
        canvas.restore()

        canvas.restore()
    }

    companion object {
        // Render compass at 280dp by default
        private const val DEFAULT_SIZE_DP = 280f
        private const val LABEL_TEXT_RATIO = 18f / 240f

        // Push labels as far from center as possible while keeping the text inside the View bounds
        private const val LABEL_RADIUS_RATIO = 0.5f - LABEL_TEXT_RATIO / 2f
    }
}
